src/main/java/com/epam/aidial/core/security/AccessService.java

```diff
@@ -33,116 +49,162 @@ public AccessService(EncryptionService encryptionService,
+    }
+
+    private Map<ResourceDescription, Set<ResourceAccessType>> lookupPermissions(
+            Set<ResourceDescription> resources, ProxyContext context, Set<ResourceAccessType> toLookup) {
+        Map<ResourceDescription, Set<ResourceAccessType>> result = new HashMap<>();
+        Set<ResourceDescription> remainingResources = new HashSet<>(resources);
+        for (PermissionRule permissionRule : permissionRules) {
+            Map<ResourceDescription, Set<ResourceAccessType>> rulePermissions =
+                    permissionRule.apply(remainingResources, context);
+
+            rulePermissions.forEach((resource, permissions) -> {
+                Set<ResourceAccessType> mergedPermissions = result.merge(resource, permissions, Sets::union);
+                if (toLookup.equals(mergedPermissions)) {
+                    remainingResources.remove(resource);
+                }
+            });
+        }
+        for (ResourceDescription resource : resources) {
+            result.computeIfAbsent(resource, r -> Set.of());
         }
 
-        return isMyResource(resource, context) || isAppResource(resource, context) || hasReviewAccess(resource, context, false);
+        return result;
     }
 
     /**
-     * Checks if USER has public access to the provided resources.
+     * Returns USER permissions to the provided public resources.
      * This method also checks admin privileges.
      *
      * @param resources - public resources
      * @param context - context
-     * @return true - if all provided resources are public and user has permissions to all of them, otherwise - false
+     * @return USER permissions
      */
-    public boolean hasPublicAccess(List<ResourceDescription> resources, ProxyContext context) {
-        return hasPublicAccess(resources, context, true);
-    }
+    private Map<ResourceDescription, Set<ResourceAccessType>> getPublicAccess(
+            Set<ResourceDescription> resources, ProxyContext context) {
+        Map<ResourceDescription, Set<ResourceAccessType>> result = new HashMap<>();
+        for (ResourceDescription resource : resources) {
+            if (!resource.isPublic()) {
+                continue;
+            }
 
-    /**
-     * Checks if USER has public access to the provided resources.
-     * This method also checks admin privileges.
-     *
-     * @param resources - public resources
-     * @param context - context
-     * @param readOnly - true to check read only access, false to check write access as well
-     * @return true - if all provided resources are public and user has permissions to all of them, otherwise - false
-     */
-    private boolean hasPublicAccess(List<ResourceDescription> resources, ProxyContext context, boolean readOnly) {
-        boolean isAllPublic = resources.stream().allMatch(ResourceDescription::isPublic);
-        if (!isAllPublic) {
-            return false;
+            if (hasAdminAccess(context)) {
```

Comment:
hasAdminAccess can be evaluated outside of the loop

Reply:
Extracted into a separate rule.